File: web/src/api/grounded-recommendation.ts
import { z } from "zod";

import { RecommendationContractError, ApiRequestError, fetchRecommendationResults, type RecommendationResultsResponse } from "./api";
import { digest, isoDate, sourceEvidence, stableId, utcDateTime } from "./trip-context";
import { parseGroundedTripInput, type GroundedTripInput } from "../features/journey/groundedTrip";
import { canonicalHash, combine, halfUp, MOOD_KEYS, nonnegative, nullableScore, ordered, same, score, without } from "./grounded-core";

export const AXIS_KEYS = ["H", "E", "R"] as const;
export const TRAIT_KEYS = ["M1", "M2", "M3", "M4", "M5", "M6"] as const;
export const CONDITION_KEYS = ["VISIT_DATE_TIME", "COMPANIONS", "TRANSPORT", "WALKING", "INDOOR_OUTDOOR", "CROWD"] as const;
const POLICY_SHA = "eceab437ed2867fa0bfe8ba7058f5648c3bbaa5c9b3891d081be1ecbcf189688";
const ids = z.array(stableId).refine(ordered);
const hashes = z.array(digest).refine(ordered);
const targets = (keys: readonly string[], nullable: boolean) => z.record(z.string(), nullable ? nullableScore : score)
  .refine((row) => same(Object.keys(row).sort(), [...keys].sort()));

const preferenceSchema = z.strictObject({
  profile_id: stableId, input_sha256: digest, projection_version: z.literal("grounded-preference-v1"),
  trip_input: z.custom<GroundedTripInput>((v) => parseGroundedTripInput(v) !== null),
  purpose: z.enum(["SIGHTSEEING", "FOOD", "LODGING", "MIXED"]),
  axis_targets: targets(AXIS_KEYS, false), trait_targets: targets(TRAIT_KEYS, true), important_traits: ids,
  condition_targets: targets(CONDITION_KEYS, true), mood_targets: targets(MOOD_KEYS, true), photo_input_sha256: digest.nullable(),
}).refine((row) => row.important_traits.every((key) => TRAIT_KEYS.includes(key as typeof TRAIT_KEYS[number])) &&
  (row.photo_input_sha256 !== null || Object.values(row.mood_targets).every((value) => value === null)));

const fitComponent = z.strictObject({
  key: stableId, expected: nullableScore, actual: nullableScore, compared: z.boolean(), difference: nullableScore,
  fit: nullableScore, weight: z.number().int().min(0).max(10_000), evidence_ids: ids,
  exclusion_reason: z.enum(["USER_UNSPECIFIED", "PLACE_UNSUPPORTED"]).nullable(),
}).refine((row) => {
  const compared = row.expected !== null && row.actual !== null;
  if (row.compared !== compared) return false;
  return compared ? row.difference === Math.abs(row.expected! - row.actual!) && row.fit === 100 - row.difference &&
    row.weight > 0 && row.evidence_ids.length > 0 && row.exclusion_reason === null
    : row.difference === null && row.fit === null && row.weight === 0 &&
      row.exclusion_reason === (row.expected === null ? "USER_UNSPECIFIED" : "PLACE_UNSUPPORTED");
});

export const fitTraceSchema = z.strictObject({
  components: z.array(fitComponent), numerator: nonnegative, denominator: nonnegative,
  score: nullableScore, compared_keys: ids,
}).refine((row) => {
  const numerator = row.components.reduce((sum, component) => sum + (component.fit ?? 0) * component.weight, 0);
  const denominator = row.components.reduce((sum, component) => sum + component.weight, 0);
  return ordered(row.components.map((component) => component.key)) && numerator === row.numerator && denominator === row.denominator &&
    row.score === (denominator ? halfUp(numerator, denominator) : null) &&
    same(row.compared_keys, row.components.filter((component) => component.compared).map((component) => component.key));
});

const dimensionSchema = z.strictObject({
  key: stableId, value: nullableScore, state: z.enum(["SUPPORTED_FACT", "SUPPORTED_INFERENCE", "UNKNOWN"]),
  evidence_ids: ids, reference_date: isoDate.nullable(), reason: z.string().min(1).max(1000),
}).refine((row) => row.state === "UNKNOWN" ? row.value === null
  : row.value !== null && row.evidence_ids.length > 0 && row.reference_date !== null);
const mismatchSchema = z.strictObject({
  axis_distance: nullableScore, trait_distance: nullableScore, raw_score: nullableScore, effective_score: nullableScore,
  compared_traits: ids, important_floor_applied: z.boolean(), state: z.enum(["INSUFFICIENT_EVIDENCE", "SUPPRESSED_LOW_CONFIDENCE",
    "NO_GUIDANCE", "GENTLE_DIFFERENCE", "MATERIAL_DIFFERENCE", "STRONG_DIFFERENCE"]), message_ko: z.string().nullable(),
});
const noveltyPair = z.strictObject({ place_id: stableId, shared_axes: z.array(z.enum(AXIS_KEYS)), shared_traits: z.array(z.enum(TRAIT_KEYS)), score });
const contributionSchema = z.strictObject({
  experience: fitTraceSchema, conditions: fitTraceSchema, traits: fitTraceSchema, mood: fitTraceSchema,
  base_relevance: score, effective_relevance: score, mood_weight: z.number().int().min(0).max(1500),
  novelty_pairs: z.array(noveltyPair), novelty_score: score, rerank_numerator: nonnegative, rerank_score: score,
}).refine((row) => {
  const base = combine([[row.experience.score, 8000], [row.conditions.score, 2000]]);
  const moodWeight = row.mood.score === null ? 0 : 1500;
  const effective = combine([[base, 10000 - moodWeight], [row.mood.score, moodWeight]]);
  const novelty = row.novelty_pairs.length ? Math.min(...row.novelty_pairs.map((pair) => pair.score)) : 0;
  const numerator = row.effective_relevance * 8500 + novelty * 1500;
  return base === row.base_relevance && moodWeight === row.mood_weight && effective === row.effective_relevance &&
    novelty === row.novelty_score && numerator === row.rerank_numerator && halfUp(numerator, 10000) === row.rerank_score;
});

export const groundedItemSchema = z.strictObject({
  rank: z.number().int().min(1).max(5), place_id: stableId, place_name_ko: z.string().min(1).max(240),
  region_code: z.string().regex(/^\d{5}$/).nullable().optional(),
  region_name: z.string().min(1).max(160).nullable().optional(),
  address_ko: z.string().min(1).max(500).nullable().optional(),
  raw_profile_sha256: digest, assessment_bundle_sha256: digest, fit_score: score, overall_confidence: nullableScore,
  axis_scores: z.array(dimensionSchema).length(3), mismatch_traits: z.array(dimensionSchema).length(6),
  contribution: contributionSchema, mismatch: mismatchSchema,
  explanations: z.array(z.strictObject({ dimension: z.enum(AXIS_KEYS), message_ko: z.string().min(1).max(200), evidence_ids: ids, reference_date: isoDate })),
  evidence: z.array(sourceEvidence), supported_axes: z.number().int().min(2).max(3), information_state: z.enum(["SUPPORTED", "LIMITED"]),
  reference_date: isoDate.nullable(), image_state: z.literal("ABSENT"),
}).refine((row) => {
  if (!same(row.axis_scores.map((axis) => axis.key), AXIS_KEYS) || !same(row.mismatch_traits.map((trait) => trait.key), TRAIT_KEYS)) return false;
  const count = row.axis_scores.filter((axis) => axis.value !== null).length;
  if (row.supported_axes !== count || row.information_state !== (count === 3 ? "SUPPORTED" : "LIMITED") || row.fit_score !== row.contribution.effective_relevance) return false;
  if (!same(Object.fromEntries(row.axis_scores.map((axis) => [axis.key, axis.value])), Object.fromEntries(row.contribution.experience.components.map((c) => [c.key, c.actual]))) ||
    !same(Object.fromEntries(row.mismatch_traits.map((trait) => [trait.key, trait.value])), Object.fromEntries(row.contribution.traits.components.map((c) => [c.key, c.actual])))) return false;
  const evidence = new Map(row.evidence.map((entry) => [entry.evidence_id, entry]));
  if (evidence.size !== row.evidence.length || row.evidence.some((entry) => entry.place_match !== null && entry.place_match.place_id !== row.place_id)) return false;
  return row.explanations.every((explanation) => {
    const axis = row.axis_scores.find((axis) => axis.key === explanation.dimension)!;
    return axis.value !== null && explanation.evidence_ids.length > 0 && explanation.evidence_ids.every((id) => axis.evidence_ids.includes(id) && evidence.has(id));
  });
});

const candidateBinding = z.strictObject({ place_id: stableId, raw_profile_sha256: digest, assessment_bundle_sha256: digest });
const authoritySchema = z.strictObject({
  kernel_version: z.literal("recommendation-kernel-v5"), config_sha256: z.literal(POLICY_SHA), release_sha256: digest,
  source_release_sha256: digest, assessment_manifest_sha256: digest, membership_sha256: digest, relation_sha256: digest,
  candidate_sha256: digest,
  candidate_assessment_sha256: digest, contextual_snapshot_sha256: hashes, photo_input_sha256: digest.nullable(),
});
const runSchema = z.strictObject({
  schema_version: z.literal("itda.grounded-recommendation-run.v1"), run_id: stableId, input_digest: digest,
  preference: preferenceSchema, authority: authoritySchema, candidate_bindings: z.array(candidateBinding), eligible_place_ids: ids,
  exclusions: z.array(z.strictObject({ place_id: stableId, reason: z.enum(["PURPOSE", "REGION", "INSUFFICIENT_SUPPORTED_AXES", "EXPLICIT_FACILITY_ABSENT", "SUPPORT_GATE"]) })),
  items: z.array(groundedItemSchema).length(5), created_at: utcDateTime, canonical_sha256: digest,
});
const resultSchema = z.strictObject({
  schema_version: z.literal("itda.grounded-recommendation-results.v1"), preference_profile_id: stableId, run: runSchema,
  release_disclosure: z.strictObject({ model: z.literal("glm-5.3-flash"), raw_release_sha256: digest, source_release_sha256: digest,
    assessment_manifest_sha256: digest, candidate_sha256: digest, config_sha256: z.literal(POLICY_SHA),
    scoring_policy: z.literal("SUPPORTED_SUBORDINATE_AGGREGATION"), image_policy: z.literal("VISUAL_MOOD_ONLY") }),
});

export type GroundedItem = z.infer<typeof groundedItemSchema>;
export type GroundedRun = z.infer<typeof runSchema>;
export type GroundedResults = z.infer<typeof resultSchema>;
export type RecommendationEnvelope = RecommendationResultsResponse | GroundedResults;
export type GroundedFitTrace = z.infer<typeof fitTraceSchema>;

function invalid(): never { throw new RecommendationContractError("INVALID_RECOMMENDATION_OUTPUT"); }

function validateMismatch(item: GroundedItem, preference: GroundedRun["preference"]): boolean {
  const trace = item.contribution, actual = item.mismatch;
  const axes = trace.experience.score === null ? null : 100 - trace.experience.score;
  const traits = trace.traits.score === null ? null : 100 - trace.traits.score;
  const raw = combine([[axes, 6500], [traits, 3500]]);
  const floor = trace.traits.components.some((c) => preference.important_traits.includes(c.key) && c.compared && c.difference !== null && c.difference >= 70);
  const effective = raw !== null && floor ? Math.max(raw, 50) : raw;
  const state = effective === null ? "INSUFFICIENT_EVIDENCE" : (item.overall_confidence ?? 0) < 65 ? "SUPPRESSED_LOW_CONFIDENCE" : effective < 30 ? "NO_GUIDANCE"
    : effective < 45 ? "GENTLE_DIFFERENCE" : effective < 60 ? "MATERIAL_DIFFERENCE" : "STRONG_DIFFERENCE";
  const message = { GENTLE_DIFFERENCE: "확인된 특성 중 기대와 조금 다른 부분이 있어요.", MATERIAL_DIFFERENCE: "확인된 특성 중 기대와 다른 부분이 있어요.", STRONG_DIFFERENCE: "기대와 다른 특성을 방문 전에 확인해 주세요." }[state as "GENTLE_DIFFERENCE"] ?? null;
  return same(actual, { axis_distance: axes, trait_distance: traits, raw_score: raw, effective_score: effective,
    compared_traits: trace.traits.compared_keys, important_floor_applied: floor, state, message_ko: message });
}

function validateNovelty(item: GroundedItem, prior: GroundedItem[]): boolean {
  if (!same(item.contribution.novelty_pairs.map((pair) => pair.place_id), prior.map((item) => item.place_id))) return false;
  const left = Object.fromEntries([...item.axis_scores, ...item.mismatch_traits].map((d) => [d.key, d.value]));
  return prior.every((other, index) => {
    const right = Object.fromEntries([...other.axis_scores, ...other.mismatch_traits].map((d) => [d.key, d.value]));
    const axes = AXIS_KEYS.filter((key) => left[key] !== null && right[key] !== null);
    const traits = TRAIT_KEYS.filter((key) => left[key] !== null && right[key] !== null);
    const distance = (keys: string[]) => keys.length ? halfUp(keys.reduce((sum, key) => sum + Math.abs(left[key]! - right[key]!), 0), keys.length) : null;
    return same(item.contribution.novelty_pairs[index], { place_id: other.place_id, shared_axes: axes, shared_traits: traits,
      score: combine([[distance(axes), 6500], [distance(traits), 3500]]) ?? 0 });
  });
}

export async function parseGroundedResults(payload: unknown, expectedRunId?: string): Promise<GroundedResults> {
  const parsed = resultSchema.safeParse(payload);
  if (!parsed.success) invalid();
  const result = parsed.data, run = result.run, authority = run.authority, disclosure = result.release_disclosure;
  if ((expectedRunId !== undefined && run.run_id !== expectedRunId) || result.preference_profile_id !== run.preference.profile_id ||
    disclosure.raw_release_sha256 !== authority.release_sha256 || disclosure.source_release_sha256 !== authority.source_release_sha256 ||
    disclosure.assessment_manifest_sha256 !== authority.assessment_manifest_sha256 || disclosure.config_sha256 !== authority.config_sha256 ||
    disclosure.candidate_sha256 !== authority.candidate_sha256 ||
    authority.photo_input_sha256 !== run.preference.photo_input_sha256) invalid();
  const bindings = new Map(run.candidate_bindings.map((row) => [row.place_id, row]));
  if (!ordered([...bindings.keys()]) || bindings.size !== run.candidate_bindings.length || !run.eligible_place_ids.every((id) => bindings.has(id)) ||
    new Set(run.items.map((item) => item.place_id)).size !== 5 ||
    new Set(run.exclusions.map((row) => row.place_id)).size !== run.exclusions.length ||
    run.exclusions.some((row) => !bindings.has(row.place_id) || run.eligible_place_ids.includes(row.place_id))) invalid();
  for (const [index, item] of run.items.entries()) {
    const binding = bindings.get(item.place_id);
    if (item.rank !== index + 1 || !run.eligible_place_ids.includes(item.place_id) || !binding ||
      (run.preference.trip_input.region_code != null && !item.region_code?.startsWith(run.preference.trip_input.region_code)) ||
      item.raw_profile_sha256 !== binding.raw_profile_sha256 || item.assessment_bundle_sha256 !== binding.assessment_bundle_sha256 ||
      !validateMismatch(item, run.preference) || !validateNovelty(item, run.items.slice(0, index))) invalid();
    for (const [trace, targets] of [[item.contribution.experience, run.preference.axis_targets], [item.contribution.traits, run.preference.trait_targets],
      [item.contribution.conditions, run.preference.condition_targets], [item.contribution.mood, run.preference.mood_targets]] as const) {
      if (!same(Object.fromEntries(trace.components.map((c) => [c.key, c.expected])), targets) ||
        trace.components.some((component) => component.weight !== (component.compared ? 1 : 0))) invalid();
    }
  }
  if (authority.candidate_assessment_sha256 !== await canonicalHash(run.candidate_bindings) ||
    run.input_digest !== await canonicalHash({ preference: run.preference, authority }) ||
    run.canonical_sha256 !== await canonicalHash(without(run, ["run_id", "created_at", "canonical_sha256"])) ||
    run.run_id !== `recommendation-run:${run.canonical_sha256.slice(0, 32)}`) invalid();
  return result;
}

export async function getGroundedJson(path: string, options: { signal?: AbortSignal; fetchImpl?: typeof fetch } = {}): Promise<unknown> {
  let response: Response;
  try { response = await (options.fetchImpl ?? fetch)(path, { signal: options.signal, credentials: "same-origin" }); }
  catch { throw new ApiRequestError("추천 정보를 불러오지 못했어요.", null); }
  let payload: unknown;
  try { payload = await response.json(); } catch { payload = null; }
  if (!response.ok) throw new ApiRequestError("추천 정보를 불러오지 못했어요.", response.status, payload);
  return payload;
}

export async function fetchRecommendationEnvelope(runId: string, options: { signal?: AbortSignal; fetchImpl?: typeof fetch } = {}): Promise<RecommendationEnvelope> {
  if (!stableId.safeParse(runId).success) invalid();
  const payload = await getGroundedJson(`/v1/recommendation-runs/${encodeURIComponent(runId)}`, options);
  if (typeof payload === "object" && payload !== null && "schema_version" in payload && payload.schema_version === "itda.grounded-recommendation-results.v1")
    return parseGroundedResults(payload, runId);
  // Reuse the frozen old decoders against the already fetched body. No second HTTP request.
  return fetchRecommendationResults(runId, { ...options, fetchImpl: async () => new Response(JSON.stringify(payload), { status: 200 }) });
}

File: backend/src/itda/domain/grounded_recommendation.py
"""Actual v5 ranking over immutable supported assessments and appearance moods."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import asdict, dataclass, field
from datetime import datetime

from itda.contracts.grounded_run import (
    AXIS_KEYS,
    CONDITION_KEYS,
    GROUNDED_POLICY,
    TRAIT_KEYS,
    GroundedAuthority,
    GroundedCandidateBinding,
    GroundedDimension,
    GroundedExclusion,
    GroundedExplanation,
    GroundedMismatchTrace,
    GroundedNoveltyPair,
    GroundedPreference,
    GroundedRecommendationItem,
    GroundedRecommendationRun,
    GroundedScoreTrace,
)
from itda.contracts.grounded_source import GroundedSourceProfile
from itda.contracts.mvp_scored_release import MvpScoredProfile
from itda.contracts.source_assessment import (
    AssessmentBundle,
    ClaimKind,
    SourceObservation,
    SupportState,
)
from itda.domain.canonical import canonical_sha256
from itda.domain.grounded_scoring import (
    combine_groups,
    fit_trace,
    half_up,
    mismatch,
    novelty,
    observed_value,
    pair_novelty,
)


class GroundedRecommendationError(ValueError):
    pass


@dataclass(frozen=True)
class GroundedCandidate:
    profile: MvpScoredProfile | GroundedSourceProfile
    assessment: AssessmentBundle
    category: str
    region_code: str | None = None
    region_name: str | None = None
    address_ko: str | None = None
    conditions: Mapping[str, SourceObservation] = field(default_factory=dict)
    moods: Mapping[str, SourceObservation] = field(default_factory=dict)
    contextual_facts: Mapping[str, SourceObservation] = field(default_factory=dict)

    @property
    def place_id(self) -> str:
        return self.profile.place_id


def _snapshot(row: SourceObservation) -> GroundedDimension:
    return GroundedDimension(
        key=row.key,
        value=observed_value(row),
        state=row.state,
        evidence_ids=tuple(sorted({e.evidence_id for e in row.evidence})),
        reference_date=row.reference_date,
        reason=row.reason,
    )


def _shared_keys(
    keys: tuple[str, ...], left: GroundedCandidate, right: GroundedCandidate
) -> tuple[str, ...]:
    return tuple(
        k
        for k in keys
        if observed_value(left.assessment.dimensions[k]) is not None
        and observed_value(right.assessment.dimensions[k]) is not None
    )


def _excluded(candidate: GroundedCandidate, preference: GroundedPreference) -> str | None:
    if preference.trip_input.region_code is not None and (
        candidate.region_code is None
        or not candidate.region_code.startswith(preference.trip_input.region_code)
    ):
        return "REGION"
    categories = {
        "SIGHTSEEING": {"관광지", "문화시설", "레포츠", "축제·공연·행사"},
        "FOOD": {"음식점"},
        "LODGING": {"숙박"},
    }
    if preference.purpose != "MIXED" and candidate.category not in categories[preference.purpose]:
        return "PURPOSE"
    if not candidate.profile.recommendation_eligible:
        return "SUPPORT_GATE"
    if (
        sum(observed_value(candidate.assessment.dimensions[a]) is not None for a in AXIS_KEYS)
        < GROUNDED_POLICY.minimum_supported_axes
    ):
        return "INSUFFICIENT_SUPPORTED_AXES"
    for requirement in preference.trip_input.required_facilities:
        fact = candidate.contextual_facts.get(
            requirement.value, candidate.assessment.facts.get(requirement.value)
        )
        if (
            fact is not None
            and fact.state == SupportState.FACT
            and fact.claim == ClaimKind.FACILITY
            and fact.value is False
        ):
            return "EXPLICIT_FACILITY_ABSENT"
    return None


def _compatible(
    candidate: GroundedCandidate,
    selected: tuple[GroundedCandidate, ...],
    forbidden: frozenset[frozenset[str]],
) -> bool:
    return all(
        candidate.profile.duplicate_group_id != p.profile.duplicate_group_id
        and frozenset({candidate.place_id, p.place_id}) not in forbidden
        for p in selected
    )


def _can_complete(
    selected: tuple[GroundedCandidate, ...],
    candidates: tuple[GroundedCandidate, ...],
    forbidden: frozenset[frozenset[str]],
) -> bool:
    needed = 5 - len(selected)
    if needed == 0:
        return True
    remaining = tuple(
        c
        for c in candidates
        if c.place_id not in {s.place_id for s in selected} and _compatible(c, selected, forbidden)
    )
    if len({c.profile.duplicate_group_id for c in remaining}) < needed:
        return False

    def search(start: int, chosen: tuple[GroundedCandidate, ...]) -> bool:
        if len(chosen) == needed:
            return True
        if len(remaining) - start < needed - len(chosen):
            return False
        return any(
            _compatible(remaining[i], chosen, forbidden) and search(i + 1, (*chosen, remaining[i]))
            for i in range(start, len(remaining))
        )

    return search(0, ())


def rank_grounded(
    *,
    candidates: tuple[GroundedCandidate, ...],
    preference: GroundedPreference,
    release_sha256: str,
    source_release_sha256: str,
    assessment_manifest_sha256: str,
    candidate_sha256: str,
    membership_sha256: str,
    relation_sha256: str,
    forbidden_pairs: tuple[tuple[str, str], ...] = (),
    contextual_snapshot_sha256: tuple[str, ...] = (),
    created_at: datetime,
) -> GroundedRecommendationRun:
    if not candidates or len({c.place_id for c in candidates}) != len(candidates):
        raise GroundedRecommendationError("INVALID_CANDIDATE_MEMBERSHIP")
    ordered = tuple(sorted(candidates, key=lambda c: c.place_id))
    for candidate in ordered:
        type(candidate.profile).model_validate_json(candidate.profile.model_dump_json())
        assessment = AssessmentBundle.model_validate_json(candidate.assessment.model_dump_json())
        if (
            assessment.place_id != candidate.place_id
            or assessment.raw_profile_sha256 != candidate.profile.profile_sha256
            or assessment.source_release_sha256 != source_release_sha256
        ):
            raise GroundedRecommendationError("ASSESSMENT_PROFILE_BINDING_INVALID")
        # Core axes must be the versioned supported-subordinate aggregate, not caller supplied.
        from itda.pipeline.grounded_assessment import aggregate_axis

        for axis in AXIS_KEYS:
            if assessment.dimensions[axis] != aggregate_axis(axis, assessment.dimensions):
                raise GroundedRecommendationError("ASSESSMENT_AXIS_AGGREGATION_INVALID")
        for key, row in candidate.conditions.items():
            SourceObservation.model_validate_json(row.model_dump_json())
            if key not in CONDITION_KEYS or row.key != key:
                raise GroundedRecommendationError("CONDITION_IDENTITY_INVALID")
            if row.state != SupportState.UNKNOWN and (
                row.claim
                not in {
                    ClaimKind.EXPERIENCE,
                    ClaimKind.FACILITY,
                    ClaimKind.OPERATING,
                    ClaimKind.CROWD,
                }
                or any(e.modality == "IMAGE_PIXELS" or e.scope != "PLACE" for e in row.evidence)
                or key == "CROWD"
                and row.claim != ClaimKind.CROWD
            ):
                raise GroundedRecommendationError("CONDITION_SOURCE_AUTHORITY_INVALID")
        for key, row in candidate.moods.items():
            SourceObservation.model_validate_json(row.model_dump_json())
            if (
                key not in preference.mood_targets
                or row.key != key
                or row.claim != ClaimKind.VISUAL_MOOD
            ):
                raise GroundedRecommendationError("MOOD_SOURCE_AUTHORITY_INVALID")
        for key, row in candidate.contextual_facts.items():
            SourceObservation.model_validate_json(row.model_dump_json())
            if (
                key != row.key
                or row.claim != ClaimKind.FACILITY
                or any(
                    e.place_match is None or e.place_match.place_id != candidate.place_id
                    for e in row.evidence
                )
            ):
                raise GroundedRecommendationError("CONTEXTUAL_FACILITY_AUTHORITY_INVALID")
    forbidden = frozenset(frozenset(pair) for pair in forbidden_pairs)
    if any(len(pair) != 2 or not pair <= {c.place_id for c in ordered} for pair in forbidden):
        raise GroundedRecommendationError("RELATION_MEMBERSHIP_INVALID")
    if forbidden_pairs != tuple(sorted(set(forbidden_pairs))) or any(
        a >= b for a, b in forbidden_pairs
    ):
        raise GroundedRecommendationError("RELATION_ORDER_INVALID")
    if relation_sha256 != canonical_sha256(
        {"place_ids": [c.place_id for c in ordered], "pairs": [list(p) for p in forbidden_pairs]}
    ):
        raise GroundedRecommendationError("RELATION_DIGEST_INVALID")
    exclusions = []
    eligible = []
    for candidate in ordered:
        reason = _excluded(candidate, preference)
        if reason:
            exclusions.append(
                GroundedExclusion.model_validate({"place_id": candidate.place_id, "reason": reason})
            )
        else:
            eligible.append(candidate)
    pool = tuple(eligible)
    if not _can_complete((), pool, forbidden):
        raise GroundedRecommendationError("INSUFFICIENT_ELIGIBLE_CANDIDATES")
    initial = {}
    for c in pool:
        experience = fit_trace(
            preference.axis_targets, {k: c.assessment.dimensions[k] for k in AXIS_KEYS}
        )
        traits = fit_trace(
            preference.trait_targets, {k: c.assessment.dimensions[k] for k in TRAIT_KEYS}
        )
        conditions = fit_trace(preference.condition_targets, c.conditions)
        mood = fit_trace(preference.mood_targets, c.moods)
        base = combine_groups(((experience.score, 8000), (conditions.score, 2000)))
        if base is None:
            raise GroundedRecommendationError("NO_COMPARABLE_EXPERIENCE")
        mood_weight = 1500 if mood.score is not None else 0
        effective = combine_groups(((base, 10000 - mood_weight), (mood.score, mood_weight)))
        if effective is None:
            raise GroundedRecommendationError("NO_COMPARABLE_EXPERIENCE")
        initial[c.place_id] = (experience, traits, conditions, mood, base, effective, mood_weight)
    selected: tuple[GroundedCandidate, ...] = ()
    items = []
    for rank in range(1, 6):
        choices = []
        for candidate in pool:
            if not _compatible(candidate, selected, forbidden):
                continue
            if not _can_complete((*selected, candidate), pool, forbidden):
                continue
            diverse = novelty(
                candidate.assessment.dimensions, tuple(p.assessment.dimensions for p in selected)
            )
            numerator = initial[candidate.place_id][5] * 8500 + diverse * 1500
            choices.append((numerator, candidate))
        if not choices:
            raise GroundedRecommendationError("INSUFFICIENT_ELIGIBLE_CANDIDATES")
        numerator, chosen = min(choices, key=lambda row: (-row[0], row[1].place_id))
        experience, traits, conditions, mood, base, effective, mood_weight = initial[
            chosen.place_id
        ]
        pair_traces = []
        for prior in selected:
            pair_traces.append(
                GroundedNoveltyPair(
                    place_id=prior.place_id,
                    shared_axes=_shared_keys(AXIS_KEYS, chosen, prior),
                    shared_traits=_shared_keys(TRAIT_KEYS, chosen, prior),
                    score=pair_novelty(chosen.assessment.dimensions, prior.assessment.dimensions),
                )
            )
        contribution = GroundedScoreTrace(
            experience=experience,
            conditions=conditions,
            traits=traits,
            mood=mood,
            base_relevance=base,
            effective_relevance=effective,
            mood_weight=mood_weight,
            novelty_pairs=tuple(pair_traces),
            novelty_score=min((p.score for p in pair_traces), default=0),
            rerank_numerator=numerator,
            rerank_score=half_up(numerator, 10000),
        )
        mismatch_value = mismatch(
            experience,
            traits,
            important_traits=frozenset(preference.important_traits),
            confidence=chosen.profile.scores.confidence
            if isinstance(chosen.profile, MvpScoredProfile)
            else 0,
        )
        messages = {
            "GENTLE_DIFFERENCE": "확인된 특성 중 기대와 조금 다른 부분이 있어요.",
            "MATERIAL_DIFFERENCE": "확인된 특성 중 기대와 다른 부분이 있어요.",
            "STRONG_DIFFERENCE": "기대와 다른 특성을 방문 전에 확인해 주세요.",
        }
        explanations = []
        for component in sorted(
            (c for c in experience.components if c.compared), key=lambda c: (-(c.fit or 0), c.key)
        )[:2]:
            row = chosen.assessment.dimensions[component.key]
            if row.reference_date is None:
                continue
            labels = {"H": "대상•원형형", "E": "의미•이미지형", "R": "자기•몰입형"}
            explanations.append(
                GroundedExplanation.model_validate(
                    {
                        "dimension": component.key,
                        "message_ko": f"{labels[component.key]}의 확인된 특성을 기대와 비교했어요.",
                        "evidence_ids": component.evidence_ids,
                        "reference_date": row.reference_date,
                    }
                )
            )
        evidence = {
            e.evidence_id: e for d in chosen.assessment.dimensions.values() for e in d.evidence
        }
        used = {eid for explanation in explanations for eid in explanation.evidence_ids}
        supported = sum(
            observed_value(chosen.assessment.dimensions[a]) is not None for a in AXIS_KEYS
        )
        items.append(
            GroundedRecommendationItem(
                rank=rank,
                place_id=chosen.place_id,
                region_code=chosen.region_code,
                region_name=chosen.region_name,
                address_ko=chosen.address_ko,
                place_name_ko=chosen.profile.place_name_ko,
                raw_profile_sha256=chosen.profile.profile_sha256,
                assessment_bundle_sha256=chosen.assessment.bundle_sha256,
                fit_score=effective,
                overall_confidence=chosen.profile.scores.confidence
                if isinstance(chosen.profile, MvpScoredProfile)
                else None,
                axis_scores=tuple(_snapshot(chosen.assessment.dimensions[a]) for a in AXIS_KEYS),
                mismatch_traits=tuple(
                    _snapshot(chosen.assessment.dimensions[t]) for t in TRAIT_KEYS
                ),
                contribution=contribution,
                mismatch=GroundedMismatchTrace(
                    **asdict(mismatch_value), message_ko=messages.get(mismatch_value.state)
                ),
                explanations=tuple(explanations),
                evidence=tuple(evidence[e] for e in sorted(used)),
                supported_axes=supported,
                information_state="SUPPORTED" if supported == 3 else "LIMITED",
                reference_date=min((e.reference_date for e in explanations), default=None),
            )
        )
        selected = (*selected, chosen)
    bindings = tuple(
        GroundedCandidateBinding(
            place_id=c.place_id,
            raw_profile_sha256=c.profile.profile_sha256,
            assessment_bundle_sha256=c.assessment.bundle_sha256,
        )
        for c in ordered
    )
    authority = GroundedAuthority(
        config_sha256=GROUNDED_POLICY.sha256,
        release_sha256=release_sha256,
        source_release_sha256=source_release_sha256,
        assessment_manifest_sha256=assessment_manifest_sha256,
        candidate_sha256=candidate_sha256,
        membership_sha256=membership_sha256,
        relation_sha256=relation_sha256,
        candidate_assessment_sha256=canonical_sha256([b.model_dump(mode="json") for b in bindings]),
        contextual_snapshot_sha256=tuple(sorted(set(contextual_snapshot_sha256))),
        photo_input_sha256=preference.photo_input_sha256,
    )
    payload = {
        "schema_version": "itda.grounded-recommendation-run.v1",
        "input_digest": canonical_sha256(
            {
                "preference": preference.model_dump(mode="json"),
                "authority": authority.model_dump(mode="json"),
            }
        ),
        "preference": preference.model_dump(mode="json"),
        "authority": authority.model_dump(mode="json"),
        "candidate_bindings": [b.model_dump(mode="json") for b in bindings],
        "eligible_place_ids": [c.place_id for c in pool],
        "exclusions": [e.model_dump(mode="json") for e in exclusions],
        "items": [i.model_dump(mode="json") for i in items],
    }
    digest = canonical_sha256(payload)
    return GroundedRecommendationRun.model_validate(
        {
            **payload,
            "run_id": f"recommendation-run:{digest[:32]}",
            "created_at": created_at,
            "canonical_sha256": digest,
        }
    )
